### PR TITLE
Kill Renderer

### PR DESCRIPTION
**Pull request by phire**
*This is a massive PR, but 90% of the changes are simply moving functions from one class/file to another*

## Background

Renderer was originally the abstraction interface for video plugins. All it's methods were virtual.

But after the video plugins were merged into the main Dolphin codebase, it slowly grew out of control.

As the backends were refactored over the years to eliminate duplicated functionality, a lot of that common code simply ended up in RenderBase. And new features would often end up added to RenderBase.

I originally started this refactor due to my work on Async Present, as that needed everything to do with presentation and the ImGui UI moved out of render. But that actually made a massive dent, and I was motivated to try and kill it completely.

What better way to stop it collecting more code over time than killing it altogether. 

### Initialization and various subobjects

Everything that Renderer owned has been moved into global g_ variables. 
Most of the initialisation logic was moved into VideoBackendBase.cpp's InitialisedShared. 

Back-ends now don't set any globals except their own.

### Presentation

There is a new `Presenter` class that handles taking finished XFBs and drawing them to screen. It also fowards the XFBs to frame dumping and is in charge of drawing the ImGui UI (the actual UI code lives in a new `OnScreenUI` class)

This class will probably be refactored further as the Async Present work continues. 

### Frame dumping and screenshots

These have moved to a new `FrameDumper` class

### Graphics API abstraction.

Almost all the virtual functions in Renderer are part of dolphin's "graphics api abstraction layer", which has slowly formed over the last decade or two.

Most of the work was done previously with the introduction of the various "AbstractX" classes, associated with texture cache cleanups and implementation of newer graphics APIs (Direct3D 12, Vulkan, Metal).
This PR simply takes the last step and yeeting these functions out of Renderer.

A new `AbstractGfx` class was created. It is now completely agnostic of any details from the flipper/hollywood GPU we are emulating, though somewhat specialised.
In the future, I hope to make it possible to have two instances of AbstractGfx. One for the main video common, and one just for presenter to render the UI over the XFB and present it to the screen.

About the only thing that actually changed in this extraction was ClearScreen. Which has been replaced with ClearRegion and the ClearScreen logic was eventually moved into FramebufferManager (where a second version called ClearEFB already existed and mostly did the same thing)

### Swap logic

This was a mess. Every time someone wanted something done once per frame, they would hack it into swap in slightly different ways.  
It really doesn't help that dolphin didn't have a clean understanding of when a frame ended.

I've introduced a new event hooking system and created new `BeforeFrame`, `AfterFrame`, `BeforePresent` and `AfterPresent` events, with clear documentation about when they should be used.

And everything that needs to do something once-per-frame now hooks those events directly. 

### Config Tracking

All the config state tracking has moved into VideoConfig.cpp. It also has a new ConfigChanged event, though I didn't wire that up everywhere. 

### EFB Scaling

This has been moved into FramebufferManager

### Widescreen heuristic 

This has been moved to a new `WidescreenManager` class. 

## Other changes

*There are probably other things that I forgot. I've been working non-stop on this refactor for a week*

### Texture cache 

The first three commits are from an older branch of mine, which got dragged along because I need them for async present. 

TextureCache now uses smart pointers to handle TCacheEntry lifecycles. This simplifies the cleanup logic and allows for more complex ownership models (such as async present being able to hold onto it's XFB independently of texture cache)

I also added a content locking feature so that things outside of texture cache can lock the contents of the texture at a given point and prevent TextureCache from overwriting it. TextureCache will do a copy-on-write instead.

## Known issues

 ~~input breaks when switching away from game window~~ pre-existing bug
 * [x] Mario Power Tennis is broken [(see fifoci)](https://fifo.ci/compare/9751448-9745643/)
 * [x] Metal backend crashes 
 * [x] XFB duplicate detection is broken. 
 * [x] Post processing seems to be broken  (including stereoscopic 3d)

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Didn't review yet, but you still have `fixup!` commits in there. Might wanna address those, or mark the PR as Draft for the time being.

**Review by BhaaLseN — COMMENTED** [review]
I guess thats all that stands out to me. And pretty much all comments at that.

The `EventHook` class looks interresting (and breaks GitHub rendering 😆 ), plus the split into essentially 3 is really neat regarding separation of what goes where.

File: Source/Core/Common/EventHook.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+using EventHook = std::unique_ptr<HookBase>;
+
+template <StringLiteral EventName, typename... CallbackArgs>
+class Event
+{
+public:
+  using CallbackType = std::function<void(CallbackArgs...)>;
+
+private:
+  struct HookImpl : public HookBase
```
Comment (iwubcode) [inline Source/Core/Common/EventHook.h:46]: `struct HookImpl final`

File: Source/Core/Common/EventHook.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+//       // Do something
+//   }, "Name of the hook");
+//
+// The hook will be automatically unregistered when the EventHook object goes out of scope.
+// Trigger events by doing:
+//
+//   MyLoveyEvent::Trigger("Hello world");
+//
+
+struct HookBase
```
Comment (iwubcode) [inline Source/Core/Common/EventHook.h:32]: [Rule of five](https://en.cppreference.com/w/cpp/language/rule_of_three)

File: Source/Core/Common/EventHook.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  static void Trigger(CallbackArgs... args)
+  {
+    for (auto& handle : m_listeners)
+      handle->m_fn(args...);
+  }
+
+private:
+  static void Remove(HookImpl* handle)
+  {
+    auto it = std::find(m_listeners.begin(), m_listeners.end(), handle);
```
Comment (iwubcode) [inline Source/Core/Common/EventHook.h:73]: Can just use `std::erase` here:  https://en.cppreference.com/w/cpp/container/vector/erase2

File: Source/Core/VideoBackends/D3D/D3DGfx.h [hunk 4]
```diff
@@ -16,11 +16,11 @@ class SwapChain;
 class DXTexture;
 class DXFramebuffer;
 
-class Renderer : public ::Renderer
+class Gfx : public ::AbstractGfx
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/D3DGfx.h:19]: `class Gfx final : public ::AbstractGfx`

File: Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp [hunk 5]
```diff
@@ -43,98 +41,75 @@ Renderer::Renderer(std::unique_ptr<SwapChain> swap_chain, float backbuffer_scale
...
-                           bool z_enable, u32 color, u32 z)
+void Gfx::ClearRegion(const MathUtil::Rectangle<int>& target_rc, bool color_enable,
+                      bool alpha_enable, bool z_enable, u32 color, u32 z)
 {
   // Use a fast path without the shader if both color/alpha are enabled.
-  const bool fast_color_clear = color_enable && (alpha_enable || !EFBHasAlphaChannel());
+  const bool fast_color_clear = color_enable && alpha_enable;
   if (fast_color_clear || z_enable)
   {
-    MathUtil::Rectangle<int> native_rc = ConvertEFBRectangle(rc);
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp:135]: I guess this was intentionally removed?

Comment (iwubcode) [reply Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp:135]: Just noticed that `D3D` didn't have a `ClearScreen()`

File: Source/Core/VideoBackends/OGL/OGLGfx.h [hunk 6]
```diff
@@ -180,7 +105,14 @@ class Renderer : public ::Renderer
...
   DepthState m_current_depth_state;
   BlendingState m_current_blend_state;
-  GLuint m_shared_read_framebuffer = 0;
-  GLuint m_shared_draw_framebuffer = 0;
+  u32 m_shared_read_framebuffer = 0;
+  u32 m_shared_draw_framebuffer = 0;
+  float m_backbuffer_scale;
 };
+
+inline OGLGfx* GetOGLGfx()
```
Comment (iwubcode) [inline Source/Core/VideoBackends/OGL/OGLGfx.h:113]: Why was this added for OGL?  I did not see it in D3D

File: Source/Core/DolphinQt/RenderWidget.cpp [hunk 7]
```diff
@@ -487,38 +487,40 @@ void RenderWidget::PassEventToImGui(const QEvent* event)
...
     if (is_down)
     {
       auto utf8 = key_event->text().toUtf8();
-      ImGui::GetIO().AddInputCharactersUTF8(utf8.constData());
+
+      if (utf8.size())
+        chars = utf8.constData();
     }
+
+    // Pass the key onto ImGui
```
[the remark this answers is not in the document]
Comment (phire) [reply Source/Core/DolphinQt/RenderWidget.cpp:501]: The problem is that this code is a bridge between Qt and ImGui. Presenter really has nothing to do with it.

However, because I was planning to move presenter (and therefor ImGui) to a new thread, I thought it would be wise to be able to gate these keys/mouse changes and only apply them once per frame.

Maybe I should beef up this abstraction. It bugged me that RenderWidget still ended up including <imgui.h> (just so it could get the constants for keys, for configuring the keymap).

I might create an intermediary API owned by presenter and convert things twice.

File: Source/Core/VideoBackends/Vulkan/VKGfx.cpp [hunk 8]
```diff
@@ -15,152 +15,98 @@
...
-
-  // Size we pass this size to vkBeginRenderPass, it has to be clamped to the framebuffer
-  // dimensions. The other backends just silently ignore this case.
-  target_rc.ClampUL(0, 0, m_target_width, m_target_height);
-
   VkRect2D target_vk_rc = {
       {target_rc.left, target_rc.top},
       {static_cast<uint32_t>(target_rc.GetWidth()), static_cast<uint32_t>(target_rc.GetHeight())}};
 
-  // Determine whether the EFB has an alpha channel. If it doesn't, we can clear the alpha
```
Comment (iwubcode) [inline Source/Core/VideoBackends/Vulkan/VKGfx.cpp:152]: Any reason this was removed?

File: Source/Core/VideoBackends/Vulkan/VKGfx.h [hunk 9]
```diff
@@ -23,19 +19,16 @@ class VKFramebuffer;
 class VKPipeline;
 class VKTexture;
 
-class Renderer : public ::Renderer
+class VKGfx : public ::AbstractGfx
```
Comment (iwubcode) [inline Source/Core/VideoBackends/Vulkan/VKGfx.h:22]: `class VKGfx final`

File: Source/Core/VideoBackends/Vulkan/VKGfx.h [hunk 10]
```diff
@@ -23,19 +19,16 @@ class VKFramebuffer;
...
 {
 public:
-  Renderer(std::unique_ptr<SwapChain> swap_chain, float backbuffer_scale);
-  ~Renderer() override;
+  VKGfx(std::unique_ptr<SwapChain> swap_chain, float backbuffer_scale);
+  ~VKGfx() override;
 
-  static Renderer* GetInstance() { return static_cast<Renderer*>(g_renderer.get()); }
+  static VKGfx* GetInstance() { return static_cast<VKGfx*>(g_gfx.get()); }
```
Comment (iwubcode) [inline Source/Core/VideoBackends/Vulkan/VKGfx.h:28]: Feels unnecessary... just use `g_gfx`?

File: Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp (hunk 5 above)
Comment (iwubcode) [reply Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp:135]: Ah, finally got to `AbstractGfx` and saw the comment :)

Comment (phire) [reply Source/Core/VideoBackends/D3D12/D3D12Gfx.cpp:135]: It's a bit misleading that github is showing the diff between these two functions, they are different functions with different APIs. `ClearScreen` takes a rect in EFB coordinates, while the new `ClearRegion` takes a target rect in native coordinate space. 

Parts of the old `ClearScreen` (such as rectangle conversion) have been moved to shared code in `FramebufferManager::ClearEFB`.

> Just noticed that D3D didn't have a ClearScreen()

Yeah, it never had a specialised version of it, relying on the common code (now in `AsbstractGfx::ClearRegion`) for all clear operations.

File: Source/Core/VideoBackends/Vulkan/VKGfx.cpp (hunk 8 above)
Comment (phire) [reply Source/Core/VideoBackends/Vulkan/VKGfx.cpp:152]: Was moved to `FramebufferManager::ClearEFB`, and now runs on all backends.  From memory there was at least one other backend with a version of this code.

File: Source/Core/VideoBackends/Vulkan/VKGfx.h (hunk 10 above)
Comment (phire) [reply Source/Core/VideoBackends/Vulkan/VKGfx.h:28]: Needs to be cast to `VKGfx` for a few places.

File: Source/Core/VideoBackends/OGL/OGLGfx.h (hunk 6 above)
Comment (phire) [reply Source/Core/VideoBackends/OGL/OGLGfx.h:113]: To remove a bunch of ugly `static_cast<OGLGfx*>(g_gfx.get());` spread throughout the codebase. Though I should probably convert it over to the `GetInstance()` static method style used elsewhere. 

Though in the medium term, both those styles are incompatible with having multiple AbstractGfxs. It can't access a global. 

Might leave this untouched until my async branch

File: Source/Core/VideoCommon/AbstractGfx.h [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+  virtual void Flush() {}
+  virtual void WaitForGPUIdle() {}
+
+  // For opengl's glDrawBuffer
+  virtual void SelectLeftBuffer() {}
+  virtual void SelectRightBuffer() {}
+  virtual void SelectMainBuffer() {}
+
+  // A simple presentation fallback, only used by video software
```
Comment (iwubcode) [inline Source/Core/VideoCommon/AbstractGfx.h:148]: This isn't a big deal but it seems a bit odd.  Should this not be handled by the presenter?

File: Source/Core/Common/StringLiteral.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2023 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <algorithm>
+
+// A useful template for passing string literals as arguments to templates
```
Comment (lioncash) [inline Source/Core/Common/StringLiteral.h:8]: This should probably be in the Common namespace

File: Source/Core/VideoCommon/AbstractGfx.h [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+  virtual void ShowImage(const AbstractTexture* source_texture,
+                         const MathUtil::Rectangle<int>& source_rc)
+  {
+  }
+
+  virtual std::unique_ptr<VideoCommon::AsyncShaderCompiler> CreateAsyncShaderCompiler();
```
Comment (iwubcode) [inline Source/Core/VideoCommon/AbstractGfx.h:154]: Was surprised this was `virtual` (and existed at all).  I will have to look at the GL implementation more later.

File: Source/Core/Common/EventHook.h [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  // Returns a handle that will unregister the listener when destroyed.
+  static EventHook Register(CallbackType callback, std::string name)
+  {
+    DEBUG_LOG_FMT(COMMON, "Registering {} handler at {} event hook", name, EventName.value);
+    auto handle = std::make_unique<HookImpl>(callback, name);
+    m_listeners.push_back(handle.get());
+    return handle;
+  }
+
+  static void Trigger(CallbackArgs... args)
```
Comment (lioncash) [inline Source/Core/Common/EventHook.h:64]: should this be 
```suggestion
  static void Trigger(const CallbackArgs&... args)
```

so that the arguments aren't potentially copied (in the event of heavyweight arguments)

File: Source/Core/VideoCommon/AbstractGfx.h (hunk 13 above)
Comment (phire) [reply Source/Core/VideoCommon/AbstractGfx.h:154]: It's probably in the wrong place. 

I'm touching all the init code (again) for my async branch. I'll make sure to deal with this then.

File: Source/Core/VideoCommon/BPStructs.cpp [hunk 15]
```diff
@@ -337,14 +341,15 @@ static void BPWritten(PixelShaderManager& pixel_shader_manager,
           false, false, yScale, s_gammaLUT[PE_copy.gamma], bpmem.triggerEFBCopy.clamp_top,
           bpmem.triggerEFBCopy.clamp_bottom, bpmem.copyfilter.GetCoefficients());
 
-      // This stays in to signal end of a "frame"
-      g_renderer->RenderToXFB(destAddr, srcRect, destStride, height, s_gammaLUT[PE_copy.gamma]);
+      // This is as closest as we have to an "end of the frame"
+      // It works 99% of the time.
```
Comment (iwubcode) [inline Source/Core/VideoCommon/BPStructs.cpp:345]: What accounts for the other 1%?

File: Source/Core/Common/EventHook.h [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+private:
+  static void Remove(HookImpl* handle)
+  {
+    auto it = std::find(m_listeners.begin(), m_listeners.end(), handle);
+    if (it != m_listeners.end())
+      m_listeners.erase(it);
+  }
+
+  inline static std::vector<HookImpl*> m_listeners = {};
```
[the remark this answers is not in the document]
Comment (phire) [reply Source/Core/Common/EventHook.h:78]: Yes. Good point.

File: Source/Core/VideoBackends/Vulkan/VKGfx.cpp (hunk 8 above)
Comment (iwubcode) [reply Source/Core/VideoBackends/Vulkan/VKGfx.cpp:152]: Because it was added to the abstract interface.  I see the pattern :)

File: Source/Core/VideoCommon/BPStructs.cpp (hunk 15 above)
Comment (phire) [reply Source/Core/VideoCommon/BPStructs.cpp:345]:
```suggestion
      // It works 99% of the time.
      // But sometimes games want to render an XFB larger than the 640x528 pixel resolution of the EFB
      // (especially when using the 3xMSAA mode, which cuts EFB resolution to 640x264). So they render
      // multiple sub-frames and copy them next to each-other in main memory. 
      // See https://dolphin-emu.org/blog/2017/11/19/hybridxfb/ for examples. 
```

File: Source/Core/VideoCommon/OnScreenUI.cpp [hunk 17]
```diff
@@ -0,0 +1,386 @@
...
+  pconfig.blending_state.srcfactor = SrcBlendFactor::SrcAlpha;
+  pconfig.blending_state.dstfactor = DstBlendFactor::InvSrcAlpha;
+  pconfig.blending_state.srcfactoralpha = SrcBlendFactor::Zero;
+  pconfig.blending_state.dstfactoralpha = DstBlendFactor::One;
+  pconfig.framebuffer_state.color_texture_format = g_presenter->GetBackbufferFormat();
+  pconfig.framebuffer_state.depth_texture_format = AbstractTextureFormat::Undefined;
+  pconfig.framebuffer_state.samples = 1;
+  pconfig.framebuffer_state.per_sample_shading = false;
+  pconfig.usage = AbstractPipelineUsage::Utility;
+  m_imgui_pipeline = g_gfx->CreatePipeline(pconfig);
```
Comment (iwubcode) [inline Source/Core/VideoCommon/OnScreenUI.cpp:166]: This is a copy from `Renderer` but it surprises me.  I didn't realize the imgui shaders went away.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 18]
```diff
@@ -740,13 +745,48 @@ void TextureCacheBase::DoLoadState(PointerWrap& p)
...
+  if (m_force_reload_textures.TestAndClear())
+  {
+    ForceReload();
+  }
+  else
+  {
+    // Flush any outstanding EFB copies to RAM, in case the game is running at an uncapped frame
+    // rate and not waiting for vblank. Otherwise, we'd end up with a huge list of pending
+    // copies.
+    g_texture_cache->FlushEFBCopies();
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp:783]: Could just be `FlushEFBCopies()`, no?

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 19]
```diff
@@ -740,13 +745,48 @@ void TextureCacheBase::DoLoadState(PointerWrap& p)
...
+  }
+
+  g_texture_cache->Cleanup(g_presenter->FrameCount());
```
Comment (iwubcode) [inline Source/Core/VideoCommon/TextureCacheBase.cpp:786]: Could just be `Cleanup(g_presenter->FrameCount())`, no?

File: Source/Core/VideoBackends/Vulkan/VKGfx.h (hunk 10 above)
Comment (iwubcode) [reply Source/Core/VideoBackends/Vulkan/VKGfx.h:28]: Ah whoops.  I don't know how I missed that.

File: Source/Core/Common/EventHook.h (hunk 14 above)
Comment (iwubcode) [reply Source/Core/Common/EventHook.h:64]: Yeah, I guess not forwarded in this case since there's a loop.  But the `CallbackType` also should be `const&` as well

File: Source/Core/Common/EventHook.h (hunk 2 above)
Comment (phire) [reply Source/Core/Common/EventHook.h]: I've deleted the copy constructor to make it clear that HookBase shouldn't be copied/moved.

File: Source/Core/Common/HookableEvent.h [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+namespace Common
+{
+struct HookBase
+{
+  virtual ~HookBase() = default;
+
+protected:
+  HookBase() = default;
+
+  // This shouldn't be copied. And since we always wrap it in unique_ptr, no need to move it either
```
Comment (iwubcode) [inline Source/Core/Common/HookableEvent.h:25]: Are you relying on the fact that specifying any of the copy-constructor/destructor removes the implicit move operations?  It might be best to specify it directly to sync with the comment:

```
HookBase(HookBase&&) = delete;
HookBase& operator=(HookBase&&) = delete;
```

File: Source/Core/Common/HookableEvent.h [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+  HookBase(const HookBase&) = delete;
+  HookBase operator=(const HookBase&) = delete;
```
Comment (iwubcode) [inline Source/Core/Common/HookableEvent.h:27]: `HookBase&` ?

File: Source/Core/Common/HookableEvent.h (hunk 20 above)
Comment (phire) [reply Source/Core/Common/HookableEvent.h:25]: I was; But it's good to be explicit.

**Comment by iwubcode** [comment]
Tested XFB/D3D11/D3D12 and a couple seconds of opengl.  I tested about 15-20 games.  Since so much is touched, I left nothing untested.  GPU decoding, screenshot saving, widescreen hack, XFBToRam, etc.

Happy to say that the crashes are now gone but post processing doesn't work at all.

File: Source/Core/DolphinQt/RenderWidget.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (phire) [reply Source/Core/DolphinQt/RenderWidget.cpp]: Done. I've created a more solid abstraction (though it's still very imgui)

**Comment by iwubcode** [comment]
Getting a warning on Opengl:  "Faield to create context for shader compiling".  I was able to get it to appear going from Vulkan (Hybrid ubershaders + compile at start) to Opengl (synchronous shaders, no compile at start).

**Comment by iwubcode** [comment]
Apologies, tested master and saw the same issue :).  So functionally I think this looks good.  Someone could test Software/Null, I did not test those (fifoci may be good enough).

**Comment by AdmiralCurtiss** [comment]
So what are we doing here, do we just yolo merge this and hope for the best?

**Comment by AdmiralCurtiss** [comment]
You still have two `!fixup` commits btw, I assume you wanted to squash those?

**Comment by phire** [comment]
Yes. There has been enough testing that we are reasonably confident.

But it touches a lot of things, so I want to wait until after we have picked a progress report release before merging. I'll squash those fixups.

**Comment by phire** [comment]
Apparently the beta build has already been locked in.

I've squashed those fixups, will merge when it passes all checks (and I double-check FIFOCI)

**Comment by phire** [comment]
Interesting. Earlier versions of this PR were causing the widescreen heuristic to be "fixed" on FIFO CI. 

But that's gone now. FIFO CI reports no changes. Guess it was triggered by one of the bugs which I fixed along the way.

**Comment by JosJuice** [comment]
Additional information: It apparently only happens with OpenGL, not with Vulkan.

**Comment by degasus** [comment]
Mssing negation in https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoBackends/OGL/OGLGfx.cpp#L137 ?

**Comment by JosJuice** [comment]
Yep, that fixes it. I'll submit a PR. Thanks!

**Comment by brujo5** [comment]
OpenGL crashes on android, vulkan has no problem
